File: HW/HW08(Java)/src/GTUSet.java
public class GTUSet<E> extends GTUContainer{
	private E data[];
	public GTUSet(){
		super();
		data=(E[])new Object[20];
	}
	public GTUSet(int size_){
		super(size_);
		data=(E[])new Object[20];
	}
	public boolean empty(){
		if(this.used_size == 0)
			return true;
		else
			return false;
	}
	public int size(){
		return this.used_size;
	}
	public int max_size(){
		return this.capacity;
	}
	public void insert(Object objec)throws java.lang.IllegalArgumentException{
		if(objec instanceof Integer || objec instanceof String || objec instanceof Double){
			E obje = ((E)objec);
			boolean control = true;
        	for(E temp_iter : data){
            	if(temp_iter == obje){
                	control = false;
            	}
        	}
        	if(control){
            	if(this.size() <= this.max_size()){
                	this.data[used_size] = obje;
            	}
            	if(this.size() > this.max_size()){
                	this.capacity = this.capacity * 2;
                	E temp[] = (E[])new Object[this.capacity];
                	int i = 0;
                	for(E temp_iter : data){
                    	temp[i] = temp_iter;
                    	i++;
                	}
                	this.data = null;
                	this.data = (E[])new Object[this.capacity];
                	this.data = temp;
                	this.data[used_size++] = obje;
            	}
            	this.used_size++;
        	}
		}
		else{
			java.lang.IllegalArgumentException e;
		}
	}
	public void erase(Object objec){
		if((objec instanceof Integer) || (objec instanceof String) || (objec instanceof Double)){
			E obje = ((E)objec);
			boolean control = false;
        	for(int i=0;i<used_size;i++){
            	if(data[i] == obje){
                	for(int k=i;k<used_size;k++){
                		data[k] = data[k+1];
                		control = true;
                	}
            	}
        	}
        	if(control)
            	used_size--;
		}
	}
	public void print()throws NullPointerException{
		for(int i=0;i<used_size;i++)
			System.out.println(data[i]);
	}
	public void clear(){
		used_size = 0;
        capacity = 1;
        E temp[] =(E[])new Object[capacity];
        data = temp;
	}
	public GTUIterator<E> begin()throws Exception{
		GTUIterator<E> iterator =new GTUIterator<E>(data);
		return iterator;
	}
	public boolean contains(Object objec){
		boolean control = false;
		if(objec instanceof Integer || objec instanceof String || objec instanceof Double){
			E obje = ((E)objec);
			for(E iter : data){
				if(obje == iter)
					control = true;
				
			}
		}
		return control;
	}
}

File: HW/HW06/GTUSet.cpp
#include "GTUSet.h"
#include <iostream>
#include <exception>
#include <memory>
using namespace std;

namespace dilaraSTL {
    template<typename T>
    GTUSet<T>::GTUSet():GTUContainer<T>(){
        shared_ptr<T> temp(new T[this->capacity]);
        this->data = temp;
    }
    template<typename T>
    bool GTUSet<T>::empty()const noexcept{
        if(this->used_size == 0)
            return true;
        else
            return false;
    }
    template<typename T>
    int GTUSet<T>::size()const noexcept{
        return this->used_size;
    }
    template<typename T>
    int GTUSet<T>::max_size()const noexcept{
        return this->capacity;
    }
    template<typename T>
    void GTUSet<T>::insert(const T &obje)noexcept(false){
        bool control = true;
        for(auto iteretor = this->begin();iteretor != this->end();++iteretor){
            if(*iteretor == obje){
                throw invalid_argument("Invalid Argument: You CANNOT insert an object in the set!");
                control = false;
            }
        }
        if(control){
            if(this->size() <= this->max_size()){
                *(this->end()) = obje;
            }
            if(this->size() > this->max_size()){
                this->capacity = this->capacity * 2;
                shared_ptr<T> temp(new T[this->capacity]);
                int i = 0;
                for(auto temp_ıter = this->begin();temp_ıter != this->end();++temp_ıter){
                    temp.get()[i] = *temp_ıter;
                    i++;
                }
                this->data = nullptr;
                this->data = temp;
                *(this->end()) = obje;
            }
            this->used_size++;
        }
    }
    template<typename T>
    void GTUSet<T>::erase(const T &obje)noexcept{
        bool control = false;
        for(auto iter=this->begin();iter!=this->end();++iter){
            if(*(iter) == obje){
                for(auto iter2 = iter; iter2 != this->end(); ++iter2){
                    *(iter2) = *(iter2.getterData()+1);
                    control = true;
                }
            }
        }
        if(control)
            this->used_size--;
    }
    template<typename T>
    void GTUSet<T>::clear()noexcept{
        this->used_size = 0;
        this->capacity = 1;
        shared_ptr<T> temp(new T[this->capacity]);
        this->data = temp;
    }
    template<typename T>
    GTUIterator<T> GTUSet<T>::begin()const noexcept{
        return (this->data).get();
    }
    template<typename T>
    GTUIterator<T> GTUSet<T>::end()const noexcept{
        return (this->data).get() + this->used_size;
    }
    
}
